File: src/main/java/com/automation/center/lighthouse/LighthouseService.java
package com.automation.center.lighthouse;

import com.automation.center.config.IConfig;
import com.automation.center.models.LighthouseResult;
import com.automation.center.models.Metric;
import com.automation.center.models.Page;
import io.qameta.allure.Step;
import io.qameta.allure.restassured.AllureRestAssured;
import io.restassured.http.ContentType;
import org.apache.http.HttpStatus;

import java.time.LocalDateTime;
import java.util.List;

import static io.restassured.RestAssured.given;

public class LighthouseService {
    private final IConfig config = IConfig.getConfig();
    private final long suiteId = config.suiteId();
    private final long reportId = createReport();
    private final List<Page> pages;
    private final List<Metric> metrics;

    public LighthouseService() {
        Pair<List<Page>, List<Metric>> pair = getPageAndMetrics();
        this.pages = pair.first();
        this.metrics = pair.second();
    }

    @Step("Create Report")
    private long createReport() {
        String endpoint = "/api/v1/report";
        var response = given()
                .filter(new AllureRestAssured())
                .contentType(ContentType.JSON)
                .body(new ReportRequest(suiteId, LocalDateTime.now().toString()))
                .when()
                .post(config.gatewayBaseUrl() + endpoint)
                .then()
                .extract()
                .jsonPath();
        return response.getLong("data.id");
    }

    @Step("Get Page and Metrics")
    private Pair<List<Page>, List<Metric>> getPageAndMetrics() {
        String endpoint = "/api/v1/suite/" + suiteId;
        var response = given()
                .filter(new AllureRestAssured())
                .when()
                .get(config.gatewayBaseUrl() + endpoint)
                .then()
                .statusCode(HttpStatus.SC_OK)
                .extract()
                .jsonPath();

        List<Page> pages = response.getList("data.pages", Page.class);
        List<Metric> metrics = response.getList("data.metrics", Metric.class);
        return new Pair<>(pages, metrics);
    }

    public void run() {
        for (Page page : pages) {
            run(page);
        }
    }

    @Step("Run Lighthouse")
    private void run(Page page) {
        var response = given()
                .filter(new AllureRestAssured())
                .queryParam("url", page.getUrl())
                .queryParam("strategy", "mobile")
                .queryParam("category", "performance")
                .when()
                .get(config.lighthouseUrl())
                .then()
                .statusCode(HttpStatus.SC_OK)
                .extract()
                .jsonPath();

        for (Metric metric : metrics) {
            Object value = response.get(metric.getJsonPath());
            LighthouseResult result = new LighthouseResult(page, metric, value.toString());
            saveResult(result);
        }
    }

    @Step("Save Result")
    private void saveResult(LighthouseResult result) {
        var endpoint = "/api/v1/result";
        given()
                .filter(new AllureRestAssured())
                .contentType(ContentType.JSON)
                .body(new ResultRequest(reportId, result.getPage().getId(), result.getMetric().getId(), result.getValue()))
                .when()
                .post(config.gatewayBaseUrl() + endpoint)
                .then()
                .statusCode(HttpStatus.SC_CREATED);
    }

    private record ReportRequest(long suiteId, String date) {
    }

    private record ResultRequest(long reportId, long pageId, long metricId, String value) {
    }

    private record Pair<F, S>(F first, S second) {
    }
}

File: lighthouse/src/main/java/com/automation/center/lighthouse/LighthouseService.kt
package com.automation.center.lighthouse

import com.automation.center.config.IConfig
import com.automation.center.models.Metric
import com.automation.center.models.Page
import com.automation.center.models.Result
import io.restassured.http.ContentType
import io.restassured.module.kotlin.extensions.Extract
import io.restassured.module.kotlin.extensions.Given
import io.restassured.module.kotlin.extensions.Then
import io.restassured.module.kotlin.extensions.When
import org.apache.http.HttpStatus
import java.time.LocalDateTime

class LighthouseService {
    private val suiteId: Long
    private val reportId: Long
    private val pages: List<Page>
    private val metrics: List<Metric>
    private val config = IConfig.getConfig()

    init {
        this.suiteId = config.suiteId()
        this.reportId = createReport()
        val (pages, metrics) = getPageAndMetrics()
        this.pages = pages
        this.metrics = metrics
    }

    private fun createReport(): Long {
        val endpoint = "/api/v1/report"
        val response = Given {
            contentType(ContentType.JSON)
            body(object {
                val suiteId = this@LighthouseService.suiteId // Use the parent class's suiteId
                val date = LocalDateTime.now().toString() // Capture current date-time
            })
        } When {
            post(config.gatewayBaseUrl() + endpoint)
        } Extract {
            jsonPath()
        }
        return response["data.id"]
    }

    private fun getPageAndMetrics(): Pair<List<Page>, List<Metric>> {
        val endpoint = "/api/v1/suite/$suiteId"
        val response = When {
            get(config.gatewayBaseUrl() + endpoint)
        } Then {
            statusCode(HttpStatus.SC_OK)
        } Extract {
            jsonPath()
        }

        val pages = response.getList("data.pages", Page::class.java)
        val metrics = response.getList("data.metrics", Metric::class.java)
        return Pair(pages, metrics)
    }

    fun run() {
        pages.forEach {
            run(it)
        }
    }

    private fun run(page: Page) {
        val response = Given {
            queryParam("url", page.url)
            queryParam("strategy", "mobile")
            queryParam("category", "performance")
        } When {
            get(config.lighthouseUrl())
        } Then {
            statusCode(HttpStatus.SC_OK)
        } Extract {
            jsonPath()
        }

        metrics.forEach() {
            val value: Any = response[it.jsonPath]
            val result = Result(page, it, value.toString())
            saveResult(result)
        }
    }

    private fun saveResult(result: Result) {
        val endpoint = "/api/v1/result"
        Given {
            contentType(ContentType.JSON)
            body(object {
                val reportId = this@LighthouseService.reportId
                val pageId = result.page.id
                val metricId = result.metric.id
                val value = result.value
            })
        } When {
            post(config.gatewayBaseUrl() + endpoint)
        } Then {
            statusCode(HttpStatus.SC_CREATED)
        }
    }
}
